tests/test_unittest/test_merge_blocks_pass.py

```diff
@@ -52,9 +52,8 @@ def test_merge_write_after_read_ij_extended(
     transform_data = merge_blocks_pass(transform_data)
     # not allowed to be merged: race condition independent of iteration order
     # write after read with extended compute domain
```

Comment:
Whoever wrote this test seems to have the opinion that this would be wrong. I don't see why since there are no offsets here. Maybe it was @DropD ?

Reply:
As the comment says: the read on `inout` happens with extended compute domain, the write does not. According to my understanding of the parallel model at the time of writing this, merging them is not allowed.

Reply:
The test came from here: https://github.com/GridTools/gt4py/issues/143#issuecomment-686455071.
Since it was written by Hannes after looking at the parallel model in-depth and confirmed by multiple people I think it requires an explanation why the stages should be allowed to merge.

Reply:
I can confirm that merging this indeed introduces a race condition in GridTools (at least if I interpret the IR here correctly, at which I have never looked so far). The second statement (`out = tmp[some ij offset]`) increases the compute extent. When fusing the last statement (`inout = in`) we get race conditions on inout due to the extended compute domain around each computational block.

Reply:
Maybe worth to mention: this race condition thus only appears if the domain is large enough, that is, larger than a single block (the default block size depends on the GT backend).

Reply:
ah, is it because these end up a single multistage with different stages, and there is no ordering guarantee on them?

Reply:
Yes, exactly. Each multistage means (approximately) one kernel call (kernel == global loop, on the GPU equivalent to a GPU kernel). The stages within the multistage are executed block-wise in this kernel. So between the stages, only block-local synchronization happens. The extreme case is a single-threaded CPU execution, where actually one block is processed after the other. Global synchronization as required in this example (because you read remote-block data from inout which is also the output field) is obviously not possible there.

Reply:
Thanks for the explanation! Off topic: I've always wondered how extended stages are implemented in GridTools. Is the CUDA grid block size covering the largest stage extent and threads are masked out (if condition) every time there's a smaller stage?

Reply:
That’s basically how it’s working yes. But there’s might be one surprise: the threads within a single block are ordered in a somewhat special way to get better alignment on memory accesses and also slightly less warp divergence. Though, not sure how much this matters on latest GPU architectures…

Reply:
@fthaler - Thanks! I came to the same conclusion as you that this pattern creates incorrect results when `tmp = inout` is in the same multistage as
```
out = tmp[1, 0, 0]
inout = in
```

When you have a moment, I have a few follow-up questions relating to this test:

```
def example(inp: Field[float], out: Field[float], inout: Field[float]):
    with computation(PARALLEL), interval(...):
        tmp = inout
        out = tmp[1, 0, 0]
        inout = inp
```

**Non-scalar temporary fields**

Naively, I would conclude that since `tmp` is not an API field (rather, it is a temporary field) that putting these into the same multistage should not incur a data race since each CUDA block could be using its own shared memory for the temporary field. At least, this is how I would try to code it in a GPU implementation.

However, when I look at the GridTools C++ generated by GT4Py, it seems like `tmp` is being defined globally instead of at the multi-stage level. For example GT v1:
```
using p_tmp = gt::tmp_arg<3, typename storage_traits<float64_t, 0, 1, 1, 1>::store_t>;
...
    computation_t gt_computation = gt::make_computation<backend_t>(
        make_grid(domain),
        gt::make_multistage(gt::execute::parallel(),
            gt::make_stage<stage__10_func>(
                p_inout(), p_tmp()
            )
        ),
        gt::make_multistage(gt::execute::parallel(),
            gt::make_stage<stage__13_func>(
                p_tmp(), p_inp(), p_inout(), p_out()
            )
        )
    );
```
and GT v2:
```
      auto GTComputationCall_111 = [](auto inout, auto inp, auto out) {
        GT_DECLARE_TMP(double, tmp);
        return multi_pass(
            execute_parallel()
                .stage(HorizontalExecution_62(), tmp, inout)
                .stage(HorizontalExecution_63(), out, tmp, inout, inp));
      };
```

Since this is at the top-level and could be used across multi-stages I assume it needs to be allocated in GPU global memory and there's no room for optimization? Or, is it still using shared memory in certain instances?

I'm wondering if there is benefit to distinguishing between temporaries that are used within a multistage and those that could be shared across them. Is GridTools may be doing some of this analysis behind the scenes?

**Stage splitting**

The `gtx86` backend executes the example correctly, but the gtc:gt:cpu_*` backends fail (off topic, but that needs to be fixed). The reason for this is that the last two statements are put in the same stage:
```
struct HorizontalExecution_63 {
  using out = inout_accessor<0, extent<0, 0, 0, 0, 0, 0>>;
  using tmp = in_accessor<1, extent<0, 1, 0, 0, 0, 0>>;
  using inout = inout_accessor<2, extent<0, 0, 0, 0, 0, 0>>;
  using inp = in_accessor<3, extent<0, 0, 0, 0, 0, 0>>;

  using param_list = make_param_list<out, tmp, inout, inp>;

  template <typename Evaluation>
  GT_FUNCTION static void
  apply(Evaluation eval, gridtools::stencil::core::interval<
                             gridtools::stencil::core::level<0, 1, 1>,
                             gridtools::stencil::core::level<1, -1, 1>>) {

    eval(out(0, 0, 0)) = eval(tmp(1, 0, 0));
    eval(inout(0, 0, 0)) = eval(inp(0, 0, 0));
  }
};
```

I'm wondering what is going on under the hood here. To me it looks like an iteration over the un-extended stencil compute domain, with an extent on tmp. Does it have to be in two stages because `tmp` was written to in an extended stage (inside a different multistage)?

**Multistage/Stage Merging Algorithm**

Evidence like this suggests that the block merging algorithm currently implemented in `master` is correctly splitting to multistages and stages to prevent race conditions. Have you identified any areas where there are unnecessary splits? Said differently, are the requirements in the code exactly what GridTools has specified, or are they overly conservative while yielding correct results?

**Outro**

If you made it this far, thanks for parsing through it :) Perhaps it would be easier to setup a meeting early next week to discuss this. Larger stencils in our dycore seem to have incorrect results with the GT backends (mostly or always combined with horizontal regions), and I suspect it's related to not fully understanding this model. We will need to fully understand to get the semantics of block merging correct with the introduction of horizontal ifs.

cc: @mbianco @anstaf